File: 3. Behavioral Design Patterns/Chain of Responsibility/Kotlin/ChainOfResponsibilityTemplate.kt
// Chain of Responsibility Pattern – Kotlin Template

abstract class Handler {
    protected var nextHandler: Handler? = null
    fun setNext(handler: Handler) { nextHandler = handler }
    abstract fun handleRequest(request: Request)
}

data class Request(val data: String = "")

class ConcreteHandlerA : Handler() {
    override fun handleRequest(request: Request) {
        when {
            canHandle(request) -> processRequest(request)
            nextHandler != null -> nextHandler!!.handleRequest(request)
        }
    }
    private fun canHandle(request: Request) = true /* condition */
    private fun processRequest(request: Request) { /* handle */ }
}

class ConcreteHandlerB : Handler() {
    override fun handleRequest(request: Request) {
        when {
            canHandle(request) -> processRequest(request)
            nextHandler != null -> nextHandler!!.handleRequest(request)
        }
    }
    private fun canHandle(request: Request) = true /* condition */
    private fun processRequest(request: Request) { /* handle */ }
}

fun main() {
    val handlerA = ConcreteHandlerA()
    val handlerB = ConcreteHandlerB()
    handlerA.setNext(handlerB)
    handlerA.handleRequest(Request())
}

File: 3. Behavioral Design Patterns/Chain of Responsibility/Typescript/ChainOfResponsibilityTemplate.ts
// Chain of Responsibility Pattern – TypeScript Template

abstract class Handler {
  protected nextHandler: Handler | null = null
  setNext(handler: Handler): void { this.nextHandler = handler }
  abstract handleRequest(request: Request): void
}

class Request { constructor(public data: string = '') {} }

class ConcreteHandlerA extends Handler {
  handleRequest(request: Request): void {
    if (this.canHandle(request)) this.processRequest(request)
    else if (this.nextHandler) this.nextHandler.handleRequest(request)
  }
  private canHandle(request: Request): boolean { return true /* condition */ }
  private processRequest(request: Request): void { /* handle */ }
}

class ConcreteHandlerB extends Handler {
  handleRequest(request: Request): void {
    if (this.canHandle(request)) this.processRequest(request)
    else if (this.nextHandler) this.nextHandler.handleRequest(request)
  }
  private canHandle(request: Request): boolean { return true /* condition */ }
  private processRequest(request: Request): void { /* handle */ }
}

;(function main() {
  const handlerA = new ConcreteHandlerA()
  const handlerB = new ConcreteHandlerB()
  handlerA.setNext(handlerB)
  handlerA.handleRequest(new Request())
})()

export {}
